src/lib.rs

```diff
@@ -1,78 +1,94 @@
-extern crate num;
+extern crate num_traits;
 extern crate rand;
 
 #[cfg(feature = "serde")]
 #[macro_use]
 extern crate serde;
 
-use num::{cast::FromPrimitive, ToPrimitive, Unsigned};
-use std::{cmp::Ord, marker::PhantomData};
-
-const WORD_SIZE: usize = 64;
+use num_traits::{cast::FromPrimitive, AsPrimitive, PrimInt, ToPrimitive, Unsigned};
+use std::{cmp::Ord, marker::PhantomData, mem::size_of};
 
 #[derive(Debug)]
 #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
-pub struct PackedVec<T> {
+pub struct PackedVec<T, StorageT = usize> {
     len: usize,
-    bits: Vec<u64>,
+    bits: Vec<StorageT>,
     item_width: usize,
     phantom: PhantomData<T>,
 }
 
-impl<'a, T> PackedVec<T>
+impl<'a, T> PackedVec<T, usize>
 where
-    T: FromPrimitive + Ord + ToPrimitive + Unsigned,
+    T: 'static + AsPrimitive<usize> + FromPrimitive + Ord + PrimInt + ToPrimitive + Unsigned,
+    usize: AsPrimitive<T>,
 {
-    /// Return a `PackedVec` containing a compressed version of the elements of
-    /// `vec`.
-    pub fn new(vec: Vec<T>) -> PackedVec<T> {
-        let len = vec.len();
-        let item_width = if let Some(v) = vec.iter().max() {
-            i_log2((*v).to_u64().unwrap())
-        } else {
+    /// Constructs a new `PackedVec` from `vec`. The `PackedVec` has `usize` backing storage, which
+    /// is likely to be the best choice in nearly all situations.
+    pub fn new(vec: Vec<T>) -> Self {
+        PackedVec::new_with_storaget(vec)
+    }
+}
+
+impl<'a, T, StorageT> PackedVec<T, StorageT>
+where
+    T: 'static + AsPrimitive<StorageT> + FromPrimitive + Ord + PrimInt + ToPrimitive + Unsigned,
+    StorageT: AsPrimitive<T> + PrimInt + Unsigned,
+{
+    /// Constructs a new `PackedVec` from `vec` (with a user-defined backing storage type).
+    pub fn new_with_storaget(vec: Vec<T>) -> PackedVec<T, StorageT> {
+        if size_of::<T>() > size_of::<StorageT>() {
+            panic!("The backing storage type must be the same size or bigger as the stored integer size.");
+        }
+        let m = vec.iter().max();
```

Comment:
I'm going to be honest: I've become allergic to long variable names as the years have rolled by. I will change it if you prefer, though, because it probably does fit in a bit better.

Reply:
I don't have a strong preference either way. The only reason I mentioned it is because I had to remind myself what `m` was when reading `let item_width = i_log2(*m.unwrap()); `. Still, I agree with you on long variable names, it will probably probably make the code less readable. Let's leave it as it is :).

Reply:
OK :)